johnzon-core/src/main/java/org/apache/johnzon/core/Snippet.java

```diff
@@ -0,0 +1,435 @@
+                try (final Buffer buffer = new Buffer()) {
+                    buffer.write(value);
+                    return buffer.get();
+                }
+            }
+        }
+    }
+
+    /**
+     * Create a serialized json representation of the supplied
+     * JsonValue, truncating the value to the specified max length.
+     * Truncated text appears with a suffix of "..."
+     *
+     * This method is thread safe.
+     *
+     * Avoid using this method in any context where there already
+     * is a JsonGeneratorFactory instance in scope. For those scenarios
+     * use the constructor that accepts a JsonGeneratorFactory instead.
+     *
+     * @param value the JsonValue to be serialized as json text
+     * @param max the maximum length of the serialized json text
+     * @return a potentially truncated json text
+     */
+    public static String of(final JsonValue value, final int max) {
+        return new Snippet(max).of(value);
+    }
+
+    /**
+     * There are several buffers involved in the creation of a json string.
+     * This class carefully manages them all.
+     *
+     * JsonGeneratorImpl with a 64k buffer (by default)
+     * ObjectStreamWriter with an 8k buffer
+     * SnippetOutputStream with a buffer of maxSnippetLength
+     *
+     * As we create json via calling the JsonGenerator it is critical we
+     * flush the work in progress all the way through these buffers and into
+     * the final SnippetOutputStream buffer.
+     *
+     * If we do not, we risk creating up to 64k of json when we may only
+     * need 50 bytes.  We could potentially optimize this code so the
+     * buffer held by JsonGeneratorImpl is also the maxSnippetLength.
+     */
+    class Buffer implements Closeable {
+        private final JsonGenerator generator;
+        private final SnippetOutputStream snippet;
+
+        private Buffer() {
+            this.snippet = new SnippetOutputStream(max);
+            this.generator = generatorFactory.createGenerator(snippet);
+        }
+
+        private void write(final JsonValue value) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            switch (value.getValueType()) {
+                case ARRAY: {
+                    write(value.asJsonArray());
+                    break;
+                }
+                case OBJECT: {
+                    write(value.asJsonObject());
+                    break;
+                }
+                default: {
+                    generator.write(value);
+                    generator.flush();
+                }
+            }
+        }
+
+        private void write(final JsonArray array) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            if (array.isEmpty()) {
+                generator.write(array);
+                generator.flush();
+                return;
+            }
+
+            generator.writeStartArray();
+            generator.flush();
```

Comment:
no need to flush *anywhere* (will not copy the comment on each line since it seems you got gluttonny ;))

Reply:
Marking as resolved based on explanations on the list